lib/supervisor/supervisor.go

```diff
@@ -0,0 +1,183 @@
+package supervisor
+
+import (
+	"errors"
+	"io/ioutil"
+	"os/exec"
+	"sync"
+
+	"github.com/andres-teleport/overseer/lib/multipipe"
+	"github.com/andres-teleport/overseer/lib/resourcecontrol"
+)
+
+var (
+	ErrUnknownJobID = errors.New("unknown job ID")
+	ErrJobFinished  = errors.New("job was already finished")
+)
+
+const (
+	StatusStarted = iota
+	StatusDone
+	StatusStopped
+)
+
+type Status struct {
+	Status   int
+	ExitCode int
+}
+
+// TODO: add option to set the environment variables
+type Job struct {
+	cmd    *exec.Cmd
+	status Status
+	stdout *multipipe.MultiPipe
+	stderr *multipipe.MultiPipe
+}
+
+type Supervisor struct {
+	mu        *sync.Mutex
+	processes map[string]*Job
+}
+
+// NewSupervisor returns a Supervisor struct that will allow starting, stopping
+// and operating with jobs
+func NewSupervisor() *Supervisor {
+	return &Supervisor{
+		mu:        &sync.Mutex{},
+		processes: make(map[string]*Job),
+	}
+}
+
+func (s *Supervisor) jobApplyFn(id string, jobFn func(*Job)) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if j, ok := s.processes[id]; ok {
+		jobFn(j)
+		return nil
+	}
+
+	return ErrUnknownJobID
+}
+
+// StartJob runs the given command and arguments, enforcing resource controls.
+// Returns a UUID to identify the job or an error on failure. cmd must be an
+// absolute path.
+func (s *Supervisor) StartJob(cmd string, args ...string) (string, error) {
+	job := &Job{
+		cmd: resourcecontrol.Command(
+			// TODO: make configurable
+			resourcecontrol.ResourceLimits{
+				CPUMax:    "10000 100000", // 10 %
+				MemMax:    "128M",
+				IOMaxRbps: "5000000",
+				IOMaxWbps: "5000000",
+			},
+			cmd,
+			args...,
+		),
+		status: Status{
+			Status: StatusStarted,
+		},
+		stdout: multipipe.NewMultiPipe(),
+		stderr: multipipe.NewMultiPipe(),
+	}
+
+	uuid, err := ioutil.ReadFile("/proc/sys/kernel/random/uuid")
+	if err != nil {
+		return "", err
+	}
+	id := string(uuid)
+
+	stdoutPipe, err := job.cmd.StdoutPipe()
+	if err != nil {
+		return "", err
+	}
+
+	stderrPipe, err := job.cmd.StderrPipe()
+	if err != nil {
+		return "", err
+	}
+
+	s.mu.Lock()
+	s.processes[id] = job
+	s.mu.Unlock()
+
+	if err := job.cmd.Start(); err != nil {
+		return "", err
+	}
+
+	stdOutDoneCh := multipipe.CopyPipe(stdoutPipe, job.stdout)
+	stdErrDoneCh := multipipe.CopyPipe(stderrPipe, job.stderr)
```

Comment:
instead of making the pipes and copying them yourself, you could probably just set `job.cmd.Stdout` and `job.cmd.Stderr` to your multipipe directly, and the `cmd` package will internally do the copying for you (https://cs.opensource.google/go/go/+/refs/tags/go1.16.7:src/os/exec/exec.go;l=303-314;drc=refs%2Ftags%2Fgo1.16.7)

Reply:
That's very good suggestion, the code looks cleaner now, thanks!